File: FindAllAnagrams.py
#Time Complexity: O(n)
#Space Complexity: O(1)

class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        if len(s)<len(p):
            return []
        res = []
        matches = 0
        m = len(s)
        n = len(p)
        d = {}
        for i in p:
            d[i] = 1+d.get(i,0)
        for i in range(m):
            if s[i] in d:
                d[s[i]]-=1
                if d[s[i]]==0:
                    matches+=1
            if i>=n:
                if s[i-n] in d:
                    d[s[i-n]]+=1
                    if d[s[i-n]]==1:
                        matches-=1
            if matches == len(d):
                res.append(i+1-n)
        return res
        

File: findAllAnagrams.cpp
//time O(s+p);
//Space O(p);

class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        //time O(s+p);
        //Space O(p);
        vector<int> result;
        unordered_map<char,int> map;
        for(int i=0;i<p.size();i++){
            char a = p[i];
            if(map.find(a)!=map.end()){
                map[a]++;
            }
            else{
                map[a]=1;
            }
        }

        int match=0;

        for(int i=0;i<s.size();i++){
            //incoming
            if(map.find(s[i])!=map.end()){
                map[s[i]]--;
                if(map[s[i]]==0){
                    match++;
                }
            }

            //outgoing

            if(i>=p.size()){
                char out = s[i-p.size()];
                if(map.find(out)!=map.end()){
                    map[out]++;
                    if(map[out]==1){
                    match--;
                }
                }
                

            }
            if (match==map.size()){
                result.push_back(i-p.size()+1);
            }
        }

        return result;

        
        
    }
};